consensus/consensus_v2.go

```diff
@@ -679,13 +681,75 @@ func (consensus *Consensus) commitBlock(blk *types.Block, committedMsg *FBFTMess
 	return nil
 }
 
+func (consensus *Consensus) rotateLeader(epoch *big.Int) {
+	prev := consensus.GetLeaderPubKey()
+	curNumber := consensus.Blockchain.CurrentHeader().Number().Uint64()
+	utils.Logger().Info().Msgf("[Rotating leader] epoch: %v rotation:%v numblocks:%d", epoch.Uint64(), consensus.Blockchain.Config().IsLeaderRotation(epoch), consensus.Blockchain.Config().LeaderRotationBlocksCount)
+	if consensus.Blockchain.Config().IsLeaderRotation(epoch) {
+		leader := consensus.GetLeaderPubKey()
+		for i := 0; i < consensus.Blockchain.Config().LeaderRotationBlocksCount; i++ {
+			header := consensus.Blockchain.GetHeaderByNumber(curNumber - uint64(i))
+			if header == nil {
+				return
+			}
+			// Previous epoch, we should not change leader.
+			if header.Epoch().Uint64() != epoch.Uint64() {
+				return
+			}
+			// Check if the same leader.
+			pub, err := chain.GetLeaderPubKeyFromCoinbase(consensus.Blockchain, header)
```

Comment:
We should keep a local cache of the leader situation for past blocks, instead of reading from the DB and parse the coinbase for every block it's pretty heavy. This DB reading and parsing should only happen when there is not enough cache data in memory to use (e.g. node first startup and for the first N blocks)

Reply:
cached